backend/msgpack/server.go

```diff
@@ -0,0 +1,326 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"math"
+	"sync"
+
+	"github.com/m3db/m3collector/backend"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	errUnrecognizedMetricType  = errors.New("unrecognized metric type")
+	errServerIsOpenOrClosed    = errors.New("server is already open or closed")
+	errServerIsNotOpenOrClosed = errors.New("server is not open or closed")
+)
+
+type serverState int
+
+const (
+	serverNotOpen serverState = iota
+	serverOpen
+	serverClosed
+)
+
+type newLockedEncoderFn func(msgpack.BufferedEncoderPool) *lockedEncoder
+
+// server partitions metrics and send them via different routes based on their partitions.
+type server struct {
+	sync.RWMutex
+
+	opts        ServerOptions
+	log         xlog.Logger
+	flushSize   int
+	encoderPool msgpack.BufferedEncoderPool
+
+	topology           topology
+	state              serverState
+	encodersByShard    []*lockedEncoder
+	newLockedEncoderFn newLockedEncoderFn
+}
+
+// NewServer creates a new server.
+func NewServer(opts ServerOptions) backend.Server {
+	s := &server{
+		opts:        opts,
+		log:         opts.InstrumentOptions().Logger(),
+		flushSize:   opts.FlushSize(),
+		encoderPool: opts.BufferedEncoderPool(),
+		topology:    newTopology(opts),
+	}
+	s.newLockedEncoderFn = newLockedEncoder
+	return s
+}
+
+func (s *server) Open() error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state != serverNotOpen {
+		return errServerIsOpenOrClosed
+	}
+	if err := s.topology.Open(); err != nil {
+		return err
+	}
+	numShards, err := s.topology.NumShards()
+	if err != nil {
+		return err
+	}
+	s.state = serverOpen
+	encoderPool := s.opts.BufferedEncoderPool()
+	encodersByShard := make([]*lockedEncoder, numShards)
+	for i := 0; i < len(encodersByShard); i++ {
+		encodersByShard[i] = s.newLockedEncoderFn(encoderPool)
+	}
+	s.encodersByShard = encodersByShard
+	return nil
+}
+
+func (s *server) WriteCounterWithPolicies(
+	id []byte,
+	val int64,
+	vp policy.VersionedPolicies,
+) error {
+	mu := unaggregated.MetricUnion{
+		Type:       unaggregated.CounterType,
+		ID:         id,
+		CounterVal: val,
+	}
+	return s.writeMetric(mu, vp)
+}
+
+func (s *server) WriteBatchTimerWithPolicies(
+	id []byte,
+	val []float64,
+	vp policy.VersionedPolicies,
+) error {
+	mu := unaggregated.MetricUnion{
+		Type:          unaggregated.BatchTimerType,
+		ID:            id,
+		BatchTimerVal: val,
+	}
+	return s.writeMetric(mu, vp)
```

Comment:
Since we're passing them by value to `s.WriteMetric` we should be alright, it should just copy the struct.